src/index.js

```diff
@@ -55,36 +61,69 @@ const create = (privateKey, value, seq, lifetime) => {
  *
  * @param {PrivateKey} privateKey - private key for signing the record.
  * @param {Uint8Array} value - value to be stored in the record.
- * @param {number} seq - number representing the current version of the record.
+ * @param {number | bigint} seq - number representing the current version of the record.
  * @param {string} expiration - expiration datetime for record in the [RFC3339]{@link https://www.ietf.org/rfc/rfc3339.txt} with nanoseconds precision.
  */
 const createWithExpiration = (privateKey, value, seq, expiration) => {
+  const expirationDate = NanoDate.fromString(expiration)
   const validityType = ipnsEntryProto.ValidityType.EOL
-  return _create(privateKey, value, seq, uint8ArrayFromString(expiration), validityType)
+
+  const ttlMs = expirationDate.toDate().getTime() - Date.now()
+  const ttlNs = (BigInt(ttlMs) * 100000n) + BigInt(expirationDate.getNano())
+
+  return _create(privateKey, value, seq, validityType, expirationDate, ttlNs)
 }
 
 /**
  * @param {PrivateKey} privateKey
  * @param {Uint8Array} value
- * @param {number} seq
- * @param {Uint8Array} isoValidity
+ * @param {number | bigint} seq
  * @param {number} validityType
+ * @param {NanoDate} expirationDate
+ * @param {bigint} ttl
  */
-const _create = async (privateKey, value, seq, isoValidity, validityType) => {
-  const signature = await sign(privateKey, value, validityType, isoValidity)
+const _create = async (privateKey, value, seq, validityType, expirationDate, ttl) => {
+  seq = BigInt(seq)
+  const isoValidity = uint8ArrayFromString(expirationDate.toString())
+  const signatureV1 = await sign(privateKey, value, validityType, isoValidity)
+  const data = createCborData(value, isoValidity, validityType, seq, ttl)
+  const sigData = ipnsEntryDataForV2Sig(data)
+  const signatureV2 = await privateKey.sign(sigData)
 
   const entry = {
     value,
-    signature: signature,
+    signature: signatureV1,
     validityType: validityType,
     validity: isoValidity,
-    sequence: seq
+    sequence: seq,
+    ttl,
+    signatureV2,
+    data
   }
 
   log(`ipns entry for ${value} created`)
   return entry
 }
 
+/**
+ * @param {Uint8Array} value
+ * @param {Uint8Array} validity
+ * @param {number} validityType
+ * @param {bigint} sequence
+ * @param {bigint} ttl
+ */
+const createCborData = (value, validity, validityType, sequence, ttl) => {
+  const data = {
+    value,
+    validity,
+    validityType,
+    sequence,
+    ttl
+  }
+
+  return cborg.encode(data)
```

Comment:
Does js cborg sorts the entries while encoding the data?

I know that in go they needed to sort manually the entries: https://github.com/ipfs/go-ipns/commit/3deb032d28934baa818e4624507dd0b1fefbbdc6#diff-3586bf1db53da12e45eb5e5072b15b3fc5ada0b80883e9945c10cead73f1ca91R76

Reply:
yes it does, and the Go ppl have to do it because @warpfork refuses to make it a codec concern or option in ipld-prime (which I will continue to moan at him about).

```
> const data = { value: new Uint8Array([1,2,3]), validity: new Uint8Array([3,4,5]), validityType: 1, sequence: 10000n, ttl: 1000000000000n }
> Buffer.from(cborg.encode(data)).toString('hex')
'a56374746c1b000000e8d4a510006576616c7565430102036873657175656e63651927106876616c6964697479430304056c76616c69646974795479706501'
```

```
$ cborg hex2diag a56374746c1b000000e8d4a510006576616c7565430102036873657175656e63651927106876616c6964697479430304056c76616c69646974795479706501
a5                                                # map(5)
  63                                              #   string(3)
    74746c                                        #     "ttl"
  1b 000000e8d4a51000                             #   uint(1000000000000)
  65                                              #   string(5)
    76616c7565                                    #     "value"
  43                                              #   bytes(3)
    010203                                        #     "\x01\x02\x03"
  68                                              #   string(8)
    73657175656e6365                              #     "sequence"
  19 2710                                         #   uint(10000)
  68                                              #   string(8)
    76616c6964697479                              #     "validity"
  43                                              #   bytes(3)
    030405                                        #     "\x03\x04\x05"
  6c                                              #   string(12)
    76616c696469747954797065                      #     "validityType"
  01                                              #   uint(1)
```

it's the RFC7049 sorting rules - length then bytewise.

it'd be great to have some fixture data to compare between implementations though to ensure this is consistent.